File: daemon.c
#include "proctitle.h"
#include "safe_rm.h"
#include <stdio.h>
#include <syslog.h>
#include <fcntl.h>
#include <dirent.h>

static volatile int sigusr1_received = 0;
static volatile int sigterm_received = 0;

// 信号处理函数
void handle_sigusr1(int sig) {
    (void)sig;
    sigusr1_received = 1;
}

void handle_sigterm(int sig) {
    (void)sig;
    sigterm_received = 1;
}

// 检查进程是否存活
int is_process_alive(pid_t pid) {
    if (kill(pid, 0) == 0) {
        return 1;
    }
    return 0;
}

// 获取PID文件路径
static char* get_pid_file_path(void) {
    static char path[MAX_PATH_LEN];
    snprintf(path, sizeof(path), "%s/%s", get_rubbish_bin_path(), DAEMON_PID_FILE);
    return path;
}

// 读取PID文件
static pid_t read_pid_file(void) {
    char *pid_file = get_pid_file_path();
    FILE *fp = fopen(pid_file, "r");
    pid_t pid = -1;
    
    if (fp != NULL) {
        if (fscanf(fp, "%d", &pid) != 1) {
            pid = -1;
        }
        fclose(fp);
    }
    
    return pid;
}

// 写入PID文件
static void write_pid_file(pid_t pid) {
    char *pid_file = get_pid_file_path();
    FILE *fp = fopen(pid_file, "w");
    
    if (fp != NULL) {
        fprintf(fp, "%d\n", pid);
        fclose(fp);
    }
}

// 删除PID文件
static void remove_pid_file(void) {
    char *pid_file = get_pid_file_path();
    unlink(pid_file);
}

// 统计A类目录
int count_a_directories(char ***dirs, int *count) {
    DIR *dp;
    struct dirent *entry;
    char *rubbish_bin = get_rubbish_bin_path();
    int capacity = 10;
    int n = 0;
    
    *dirs = malloc(capacity * sizeof(char*));
    if (*dirs == NULL) {
        return -1;
    }
    
    dp = opendir(rubbish_bin);
    if (dp == NULL) {
        free(*dirs);
        return -1;
    }
    
    while ((entry = readdir(dp)) != NULL) {
        if (entry->d_type == DT_DIR || entry->d_type == DT_UNKNOWN) {
            if (strcmp(entry->d_name, ".") == 0 || 
                strcmp(entry->d_name, "..") == 0 ||
                strcmp(entry->d_name, OLD_DIR) == 0) {
                continue;
            }
            
            if (is_valid_timestamp_format(entry->d_name)) {
                if (n >= capacity) {
                    capacity *= 2;
                    char **new_dirs = realloc(*dirs, capacity * sizeof(char*));
                    if (new_dirs == NULL) {
                        break;
                    }
                    *dirs = new_dirs;
                }
                
                (*dirs)[n] = malloc(strlen(entry->d_name) + 1);
                if ((*dirs)[n] != NULL) {
                    strcpy((*dirs)[n], entry->d_name);
                    n++;
                }
            }
        }
    }
    
    closedir(dp);
    *count = n;
    return 0;
}

// 比较时间戳（用于排序）
static int compare_timestamp(const void *a, const void *b) {
    return strcmp(*(const char**)a, *(const char**)b);
}

// 递归删除目录（压缩成功后清理）
static void remove_directory_recursive(const char *path) {
    DIR *dp;
    struct dirent *entry;
    struct stat st;
    char full_path[MAX_PATH_LEN];
    
    dp = opendir(path);
    if (dp == NULL) {
        return;
    }
    
    while ((entry = readdir(dp)) != NULL) {
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0) {
            continue;
        }
        
        snprintf(full_path, sizeof(full_path), "%s/%s", path, entry->d_name);
        if (stat(full_path, &st) == 0) {
            if (S_ISDIR(st.st_mode)) {
                remove_directory_recursive(full_path);
            } else {
                unlink(full_path);
            }
        }
    }
    
    closedir(dp);
    rmdir(path);
}

// 压缩旧目录
int compress_old_directories(void) {
    char **dirs = NULL;
    int count = 0;
    int i;
    int compressed = 0;
    char *rubbish_bin = get_rubbish_bin_path();
    char old_dir[MAX_PATH_LEN];
    
    snprintf(old_dir, sizeof(old_dir), "%s/%s", rubbish_bin, OLD_DIR);
    
    // 获取所有A类目录
    if (count_a_directories(&dirs, &count) != 0) {
        return -1;
    }
    
    if (count <= COMPRESS_THRESHOLD) {
        // 目录数量不足，不压缩
        for (i = 0; i < count; i++) {
            free(dirs[i]);
        }
        free(dirs);
        return 0;
    }
    
    // 按时间排序（老的在前）
    qsort(dirs, count, sizeof(char*), compare_timestamp);
    
    // 检查最老的目录是否超过压缩天数
    char oldest_dir[MAX_PATH_LEN];
    snprintf(oldest_dir, sizeof(oldest_dir), "%s/%s", rubbish_bin, dirs[0]);
    int days = days_since_creation(oldest_dir);
    
    if (days < COMPRESS_DAYS) {
        // 最老的目录也不够旧，不压缩
        for (i = 0; i < count; i++) {
            free(dirs[i]);
        }
        free(dirs);
        return 0;
    }
    
    // 确保old目录存在
    create_directory_recursive(old_dir);
    
    // 按COMPRESS_THRESHOLD个一组压缩（从老的开始）
    // 保留最新的 COMPRESS_THRESHOLD 个目录不压缩
    // 可压缩的目录数量 = 总数 - 保留数量
    int dirs_to_compress = count - COMPRESS_THRESHOLD;
    
    for (i = 0; i < dirs_to_compress; i += COMPRESS_THRESHOLD) {
        char tar_file[MAX_PATH_LEN];
        char timestamp[FULL_TIME_LEN];
        int j;
        
        // 生成压缩包时间戳
        get_current_timestamp(timestamp, sizeof(timestamp));
        snprintf(tar_file, sizeof(tar_file), "%s/%s.tar.gz", old_dir, timestamp);
        
        // 构建完整路径数组
        char **paths = malloc(COMPRESS_THRESHOLD * sizeof(char*));
        if (paths == NULL) {
            continue;
        }

        int path_count = 0;
        for (j = 0; j < COMPRESS_THRESHOLD && (i + j) < dirs_to_compress; j++) {
            // 构建完整路径: rubbish_bin/dirs[i+j]
            size_t len = strlen(rubbish_bin) + strlen(dirs[i + j]) + 2;
            paths[j] = malloc(len);
            if (paths[j]) {
                snprintf(paths[j], len, "%s/%s", rubbish_bin, dirs[i + j]);
                path_count++;
            }
        }

        // 使用 tarlib 创建 tar.gz（压缩级别 6，默认）
        int retval = tgz_create(tar_file, (const char **)paths, path_count, 6);
        
        if (retval == 0) {
            // 压缩成功，删除原目录
            for (j = 0; j < path_count; j++) {
                remove_directory_recursive(paths[j]);
            }
            compressed++;
        }

        // 释放路径内存
        for (j = 0; j < path_count; j++) {
            free(paths[j]);
        }
        free(paths);
    }
    
    for (i = 0; i < count; i++) {
        free(dirs[i]);
    }
    free(dirs);
    
    return compressed;
}

// 清理旧压缩包
int cleanup_old_archives(void) {
    DIR *dp;
    struct dirent *entry;
    char *rubbish_bin = get_rubbish_bin_path();
    char old_dir[MAX_PATH_LEN];
    int deleted = 0;
    regex_t regex;
    
    snprintf(old_dir, sizeof(old_dir), "%s/%s", rubbish_bin, OLD_DIR);
    
    dp = opendir(old_dir);
    if (dp == NULL) {
        return 0;
    }
    
    // 编译正则表达式匹配时间戳格式
    regcomp(&regex, "^[0-9]{4}-[0-9]{2}-[0-9]{2}-[0-9]{2}:[0-9]{2}:[0-9]{2}\\.tar\\.gz$", REG_EXTENDED);
    
    while ((entry = readdir(dp)) != NULL) {
        if (regexec(&regex, entry->d_name, 0, NULL, 0) == 0) {
            char file_path[MAX_PATH_LEN];
            struct stat st;
            time_t now;
            double diff;
            
            snprintf(file_path, sizeof(file_path), "%s/%s", old_dir, entry->d_name);
            
            if (stat(file_path, &st) == 0) {
                now = time(NULL);
                diff = difftime(now, st.st_mtime);
                
                if (diff > CLEANUP_DAYS * 24 * 3600) {
                    // 超过60天，删除
                    if (unlink(file_path) == 0) {
                        deleted++;
                    }
                }
            }
        }
    }
    
    regfree(&regex);
    closedir(dp);
    
    return deleted;
}

// 守护进程主函数
void daemon_main(void) {
    struct sigaction sa;
    
    // 设置信号处理
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = handle_sigusr1;
    sigaction(SIGUSR1, &sa, NULL);
    
    sa.sa_handler = handle_sigterm;
    sigaction(SIGTERM, &sa, NULL);
    sigaction(SIGINT, &sa, NULL);
    
    // 忽略SIGCHLD防止僵尸进程
    signal(SIGCHLD, SIG_IGN);
    
    // 写入PID文件
    write_pid_file(getpid());
    
    // 主循环
    while (!sigterm_received) {
        if (sigusr1_received) {
            sigusr1_received = 0;
            
            // 调试日志
            FILE *logfp = fopen("/data/data/com.termux/files/home/.rubbishbin/.daemon.log", "a");
            if (logfp) {
                fprintf(logfp, "[trashd] Received SIGUSR1, starting maintenance\n");
                fclose(logfp);
            }
            
            // 执行维护任务
            int compressed = compress_old_directories();
            int deleted = cleanup_old_archives();
            
            // 记录日志
            openlog("safe_rm_daemon", LOG_PID, LOG_USER);
            syslog(LOG_INFO, "Maintenance: compressed %d groups, deleted %d old archives", 
                   compressed, deleted);
            closelog();
        }
        
        // 睡眠等待信号
        sleep(1);
    }
    
    // 清理
    remove_pid_file();
}

// 启动守护进程
int start_daemon(void) {
    pid_t pid;
    
    // 创建子进程
    pid = fork();
    if (pid == -1) {
        return -1;
    }
    
    if (pid > 0) {
        // 父进程退出
        return 0;
    }
    
    // 子进程
    
    // 创建新会话
    if (setsid() == -1) {
        exit(1);
    }
    
    // 再次fork防止重新获取控制终端
    pid = fork();
    if (pid == -1) {
        exit(1);
    }
    
    if (pid > 0) {
        // 第一个子进程退出
        exit(0);
    }
    
    // 守护进程
    proctitle_init_global();
    proctitle_set("trashd");
    
    // 更改工作目录
    chdir("/");
    
    // 关闭标准文件描述符
    close(STDIN_FILENO);
    close(STDOUT_FILENO);
    close(STDERR_FILENO);
    
    // 守护进程主循环
    daemon_main();
    
    exit(0);
}

// 获取PID文件锁（非阻塞）
// 返回值：锁定的文件描述符，如果获取失败返回 -1
static int lock_pid_file(void) {
    char *pid_file = get_pid_file_path();
    int fd = open(pid_file, O_RDWR | O_CREAT, 0644);
    if (fd < 0) {
        return -1;
    }
    
    // 尝试获取排他锁（非阻塞）
    struct flock fl;
    memset(&fl, 0, sizeof(fl));
    fl.l_type = F_WRLCK;
    fl.l_whence = SEEK_SET;
    fl.l_start = 0;
    fl.l_len = 0;
    
    if (fcntl(fd, F_SETLK, &fl) < 0) {
        // 获取锁失败，可能有其他进程正在启动守护进程
        close(fd);
        return -1;
    }
    
    return fd;
}

// 检查并触发守护进程（带文件锁，确保单例）
int check_and_trigger_daemon(void) {
    // 首先尝试获取文件锁
    int lock_fd = lock_pid_file();
    if (lock_fd < 0) {
        // 获取锁失败，说明有其他进程正在启动守护进程
        // 短暂等待后再次检查
        usleep(100000);  // 100ms
        pid_t pid = read_pid_file();
        if (pid > 0 && is_process_alive(pid)) {
            kill(pid, SIGUSR1);
        }
        return 0;
    }
    
    // 获得锁后，再次检查PID文件（双重检查锁定模式）
    pid_t pid = read_pid_file();
    
    if (pid > 0 && is_process_alive(pid)) {
        // 守护进程已运行，发送SIGUSR1触发维护
        kill(pid, SIGUSR1);
        close(lock_fd);
        return 0;
    }
    
    // 清理失效的PID文件
    if (pid > 0) {
        remove_pid_file();
        // 重新打开文件以获取新的文件描述符
        close(lock_fd);
        lock_fd = lock_pid_file();
        if (lock_fd < 0) {
            return -1;
        }
    }
    
    // 启动新的守护进程
    int result = start_daemon();
    
    // 关闭锁文件描述符
    close(lock_fd);
    
    return result;
}

File: src/daemon.rs
//! trashd - SafeRm daemon for automatic maintenance

use anyhow::Result;
use nix::fcntl::{fcntl, FcntlArg, OFlag};
use nix::sys::stat::Mode;
use nix::unistd::{fork, ForkResult, Pid, execvp, setsid};
use nix::sys::signal::{kill, Signal, SigSet, sigprocmask};
use nix::sys::signalfd::SignalFd;
use libc::nice;
use std::fs;
use std::env;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::SystemTime;
use std::ffi::CString;
use std::os::unix::io::RawFd;
use std::os::unix::fs::MetadataExt;
use std::os::unix::ffi::OsStrExt;
use chrono::Local;
use crate::{
    create_dir_all, get_rubbish_bin_path, get_timestamp, 
    parse_timestamp, CLEANUP_DAYS, COMPRESS_DAYS, 
    COMPRESS_THRESHOLD, OLD_DIR, DAEMON_NAME, DAEMON_ENV_VAR, 
    DAEMON_PID_FILE, DAEMON_LOG_FILE, get_daemon_dir,
    ARCHIVE_CAPACITY,
};

const MAX_LOG_SIZE: u64      = 10 * 1024 * 1024;
const MAX_LOG_FILES: u32     = 3;
const COMPRESSION_LEVEL: u32 = 6;

/// Get PID file path
pub fn get_pid_file_path() -> Result<PathBuf> {
    let daemon_dir = get_daemon_dir()?;
    Ok(daemon_dir.join(DAEMON_PID_FILE))
}

/// Read PID from PID file
pub fn read_pid_file() -> Option<Pid> {
    let pid_file = get_pid_file_path().ok()?;
    let content = fs::read_to_string(&pid_file).ok()?;
    let pid: i32 = content.trim().parse().ok()?;
    Some(Pid::from_raw(pid))
}

/// Write PID to PID file
pub fn write_pid_file(pid: Pid) -> Result<()> {
    let pid_file = get_pid_file_path()?;
    fs::write(&pid_file, pid.as_raw().to_string())?;
    Ok(())
}

/// Remove PID file
pub fn remove_pid_file() -> Result<()> {
    let pid_file = get_pid_file_path()?;
    if pid_file.exists() {
        fs::remove_file(&pid_file)?;
    }
    Ok(())
}

/// Check if process is alive
pub fn is_process_alive(pid: Pid) -> bool {
    kill(pid, None).is_ok()
}

/// Lock PID file (non-blocking)
/// Returns file descriptor if successful, None otherwise
pub fn lock_pid_file() -> Option<RawFd> {
    let pid_file = get_pid_file_path().ok()?;
    let fd = nix::fcntl::open(
        &pid_file,
        OFlag::O_RDWR | OFlag::O_CREAT,
        Mode::from_bits_truncate(0o644),
    )
    .ok()?;

    // Try to get exclusive lock (non-blocking)
    let flock = libc::flock {
        l_type: libc::F_WRLCK as i16,
        l_whence: libc::SEEK_SET as i16,
        l_start: 0,
        l_len: 0,
        l_pid: 0,
    };

    if fcntl(fd, FcntlArg::F_SETLK(&flock)).is_err() {
    let _ = nix::unistd::close(fd);
        if let Some(pid) = read_pid_file() {
            if !is_process_alive(pid) {
                return lock_pid_file();
            }
        }
        return None;
    }

    Some(fd)
}

/// Rotate the daemon log file if it exceeds MAX_LOG_SIZE
/// Old logs are kept as daemon.log.1, daemon.log.2, etc.
fn rotate_log_if_needed(daemon_dir: &Path) {
    let log_path = daemon_dir.join(DAEMON_LOG_FILE);

    if let Ok(metadata) = fs::metadata(&log_path) {
        if metadata.len() < MAX_LOG_SIZE {
            return;
        }
    } else {
        return;
    }

    // Rename existing historical logs: .2 → .3, .1 → .2
    for i in (1..MAX_LOG_FILES).rev() {
        let old = daemon_dir.join(format!("{}.{}", DAEMON_LOG_FILE, i));
        let new = daemon_dir.join(format!("{}.{}", DAEMON_LOG_FILE, i + 1));
        let _ = fs::rename(&old, &new);
    }

    // Rename current log to .1
    let backup = daemon_dir.join(format!("{}.1", DAEMON_LOG_FILE));
    let _ = fs::rename(&log_path, &backup);
}

/// Unlock and close PID file
pub fn unlock_pid_file(fd: RawFd) {
    let _ = nix::unistd::close(fd);
}

/// Count A directories (timestamp directories)
pub fn count_a_directories(rubbish_bin: &Path) -> Result<Vec<String>> {
    let mut dirs = Vec::new();

    if !rubbish_bin.exists() {
        return Ok(dirs);
    }

    for entry in fs::read_dir(rubbish_bin)? {
        let entry = entry?;
        let path = entry.path();
        let name = entry.file_name().to_string_lossy().to_string();

        if name == OLD_DIR || name.starts_with('.') {
            continue;
        }

        if path.is_dir() && parse_timestamp(&name).is_some() {
            dirs.push(name);
        }
    }

    Ok(dirs)
}

/// Calculate days since directory creation
pub fn days_since_creation(metadata: &fs::Metadata) -> Result<i64> {
    let modified = metadata.modified()?;
    let now = SystemTime::now();
    let duration = now.duration_since(modified).unwrap_or_default();
    Ok(duration.as_secs() as i64 / (24 * 3600))
}

/// Remove directory recursively
pub fn remove_directory_recursive(path: &Path) -> Result<()> {
    if path.is_dir() {
        fs::remove_dir_all(path)?;
    } else {
        fs::remove_file(path)?;
    }
    Ok(())
}

/// Create tar.gz archive from directories
pub fn create_tar_gz(archive_path: &Path, dirs: &[&Path], compression_level: u32) -> Result<()> {
    use flate2::write::GzEncoder;
    use flate2::Compression;
    use tar::Builder;

    let file = fs::File::create(archive_path)?;
    let encoder = GzEncoder::new(file, Compression::new(compression_level));
    let mut builder = Builder::new(encoder);

    for dir in dirs {
        builder.append_dir_all(
            dir.file_name().unwrap_or_default().to_string_lossy().as_ref(),
            dir,
        )?;
    }

    builder.finish()?;
    Ok(())
}

/// Compress old directories
pub fn compress_old_directories() -> Result<usize> {
    let rubbish_bin = get_rubbish_bin_path()?;
    let old_dir = rubbish_bin.join(OLD_DIR);

    let mut dirs = count_a_directories(&rubbish_bin)?;

    if dirs.len() <= COMPRESS_THRESHOLD {
        return Ok(0);
    }

    // Sort by timestamp (oldest first)
    dirs.sort();

    // Check if oldest directory is old enough
    let oldest_dir = rubbish_bin.join(&dirs[0]);
    let oldest_metadata = fs::metadata(&oldest_dir)?;   // 仅一次 stat
    let days = days_since_creation(&oldest_metadata)?;  // 传入元数据

    if days < COMPRESS_DAYS {
        return Ok(0);
    }

    // Ensure old directory exists
    create_dir_all(&old_dir)?;

    let dirs_to_compress = dirs.len() - COMPRESS_THRESHOLD;
    let full_groups = dirs_to_compress / ARCHIVE_CAPACITY;
    let mut compressed = 0;
    let mut group_index = 0;

    // Compress only full groups (each of ARCHIVE_CAPACITY directories)
    for i in 0..full_groups {
        let timestamp = get_timestamp();
        let tar_file = if group_index == 0 {
            old_dir.join(format!("{}.tar.gz", timestamp))
        } else {
            old_dir.join(format!("{}_{}.tar.gz", timestamp, group_index))
        };
        group_index += 1;

        let start = i * ARCHIVE_CAPACITY;
        let group: Vec<PathBuf> = dirs[start..start + ARCHIVE_CAPACITY]
            .iter()
            .map(|name| rubbish_bin.join(name))
            .collect();

        let group_refs: Vec<&Path> = group.iter().map(|p| p.as_path()).collect();

        if create_tar_gz(&tar_file, &group_refs, COMPRESSION_LEVEL).is_ok() {
            // Generate manifest BEFORE removing original directories
            let _ = crate::generate_manifest(&tar_file, &group_refs);
            // Remove original directories after successful compression
            for dir in &group {
                let _ = remove_directory_recursive(dir);
            }
            compressed += 1;
        }
    }

    Ok(compressed)
}

/// Clean up old archives
pub fn cleanup_old_archives() -> Result<usize> {
    let rubbish_bin = get_rubbish_bin_path()?;
    let old_dir = rubbish_bin.join(OLD_DIR);
    let mut deleted = 0;

    if !old_dir.exists() {
        return Ok(0);
    }

    let regex =
        regex::Regex::new(r"^\d{4}-\d{2}-\d{2}-\d{2}:\d{2}:\d{2}(?:_\d+)?\.tar\.gz$")
            .unwrap();

    for entry in fs::read_dir(&old_dir)? {
        let entry = entry?;
        let name = entry.file_name().to_string_lossy().to_string();

        if regex.is_match(&name) {
            let metadata = entry.metadata()?;
            let modified = metadata.modified()?;
            let now = SystemTime::now();
            let duration = now.duration_since(modified).unwrap_or_default();
            let days = duration.as_secs() as i64 / (24 * 3600);

            if days > CLEANUP_DAYS {
                let path = entry.path();
                if fs::remove_file(&path).is_ok() {
                    // Also remove manifest if exists
                    let manifest_path = crate::manifest_path(&path);
                    let _ = fs::remove_file(manifest_path);
                    deleted += 1;
                }
            }
        }
    }

    Ok(deleted)
}

/// Perform maintenance tasks.
/// Returns (compressed_groups, fully_deleted_archives, expired_archives).
pub fn perform_maintenance() -> Result<(usize, usize, usize)> {
    let compressed = compress_old_directories()?;
    let fully_deleted = crate::cleanup_fully_deleted_archives()?;
    let expired = cleanup_old_archives()?;

    // Only refresh cache if something actually changed
    if compressed > 0 || fully_deleted > 0 || expired > 0 {
        let _ = crate::refresh_cache();
    }

    Ok((compressed, fully_deleted, expired))
}

/// Check and trigger daemon
pub fn check_and_trigger_daemon() -> Result<()> {
    // Try to lock PID file
    let lock_fd = match lock_pid_file() {
        Some(fd) => fd,
        None => {
            // Another process is starting daemon, wait and signal
            std::thread::sleep(std::time::Duration::from_millis(100));
            if let Some(pid) = read_pid_file() {
                if is_process_alive(pid) {
                    let _ = kill(pid, Signal::SIGUSR1);
                }
            }
            return Ok(());
        }
    };

    // Double-check after acquiring lock
    if let Some(pid) = read_pid_file() {
        if is_process_alive(pid) {
            // Daemon already running, signal it
            let _ = kill(pid, Signal::SIGUSR1);
            unlock_pid_file(lock_fd);
            return Ok(());
        }
    }

    // Clean up stale PID file
    let _ = remove_pid_file();

    // Start new daemon
    let result = start_daemon();

    unlock_pid_file(lock_fd);
    result
}

/// Start the trash daemon (using double fork + exec method).
pub fn start_daemon() -> Result<()> {
    let current_exe = env::current_exe()?;
    let exe_path = CString::new(current_exe.as_os_str().as_bytes())?;

    match unsafe { fork()? } {
        ForkResult::Parent { .. } => Ok(()),
        ForkResult::Child => {
            setsid()?;

            match unsafe { fork()? } {
                ForkResult::Parent { .. } => std::process::exit(0),
                ForkResult::Child => {
                    // Standard daemon settings
                    env::set_current_dir("/").ok();
                    let _ = nix::unistd::close(0);
                    let _ = nix::unistd::close(1);
                    let _ = nix::unistd::close(2);

                    // Set environment variables to identify daemon mode
                    env::set_var(DAEMON_ENV_VAR, "1");

                    // Construct argv: [DAEMON_NAME, "..."]
                    // The second parameter can be anything
                    let argv0 = CString::new(DAEMON_NAME)?;
                    let argv1 = CString::new("Manager")?;  // Placeholder
                    let args = [&argv0, &argv1];

                    execvp(&exe_path, &args)?;
                    // execvp doesn't return when successful
                    unreachable!();
                }
            }
        }
    }
}

fn is_same_file(file: &std::fs::File, path: &Path) -> bool {
    if let (Ok(f_meta), Ok(p_meta)) = (file.metadata(), std::fs::metadata(path)) {
        let f_ino = (f_meta.ino(), f_meta.dev());
        let p_ino = (p_meta.ino(), p_meta.dev());
        f_ino == p_ino
    } else {
        false
    }
}

/// Daemon main loop – event‑driven via signalfd (no polling).
fn daemon_main() {
    // Lower CPU priority to avoid affecting foreground operations
    let _ = unsafe { nice(10) };

    // Create the set of signals we want to handle.
    let mut mask = SigSet::empty();
    mask.add(Signal::SIGUSR1);
    mask.add(Signal::SIGTERM);
    mask.add(Signal::SIGINT);

    // Block the signals so they are delivered through signalfd instead.
    sigprocmask(
        nix::sys::signal::SigmaskHow::SIG_BLOCK,
        Some(&mask),
        None,
    )
    .expect("Failed to block signals");

    // Create a signalfd that will wake up when any blocked signal arrives.
    let sfd = SignalFd::new(&mask).expect("Failed to create signalfd");

    let daemon_dir = get_daemon_dir().expect("Failed to get daemon dir");

    let log_path = daemon_dir.join(DAEMON_LOG_FILE);
    let mut log_writer = std::io::BufWriter::new(
        std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&log_path)
            .expect("Failed to open daemon log"),
    );

    // Write PID file (existing logic).
    let _ = write_pid_file(nix::unistd::getpid());

    // Main event loop.
    loop {
        match sfd.read_signal() {
            Ok(Some(siginfo)) => {
                let signo = siginfo.ssi_signo as i32;
                let signal = Signal::try_from(signo).unwrap_or(Signal::SIGUSR1);

                if signal == Signal::SIGUSR1 {
                    rotate_log_if_needed(&daemon_dir);

                    if !is_same_file(log_writer.get_ref(), &log_path) {
                        log_writer = std::io::BufWriter::new(
                            std::fs::OpenOptions::new()
                                .create(true)
                                .append(true)
                                .open(&log_path)
                                .expect("Failed to reopen log after rotation"),
                        );
                    }

                    let msg = format!(
                        "[{} {}] Received SIGUSR1, starting maintenance\n",
                        DAEMON_NAME,
                        Local::now().format("%Y-%m-%d-%H:%M:%S")
                    );
                    let _ = log_writer.write_all(msg.as_bytes());
                    let _ = log_writer.flush();

                    let (compressed, fully_deleted, expired) =
                        perform_maintenance().unwrap_or((0, 0, 0));

                    let msg = format!(
                        "[{} {}] Maintenance: compressed {} groups, cleaned {} fully-deleted archives, deleted {} expired archives\n",
                        DAEMON_NAME,
                        Local::now().format("%Y-%m-%d-%H:%M:%S"),
                        compressed,
                        fully_deleted,
                        expired
                    );
                    let _ = log_writer.write_all(msg.as_bytes());
                    let _ = log_writer.flush();
                } else if signal == Signal::SIGTERM || signal == Signal::SIGINT {
                    // Termination signal – exit the loop.
                    break;
                }
            }
            Ok(None) => {
                // Should not happen; sleep briefly to avoid a busy loop just in case.
                std::thread::sleep(std::time::Duration::from_millis(100));
            }
            Err(e) => {
                // signalfd read error – log and retry after a short delay.
                eprintln!("trashd: signalfd read error: {}", e);
                std::thread::sleep(std::time::Duration::from_secs(1));
            }
        }
    }

    // Clean up PID file before exiting.
    let _ = remove_pid_file();
}

/// Run the main loop of the daemon directly
pub fn daemon_main_direct() {
    daemon_main();
}


